src/gc/impl/conservative/gc.d

```diff
@@ -1003,68 +1029,64 @@ struct GC
     }
 
 
-    /**
-     * Verify that pointer p:
-     *  1) belongs to this memory pool
-     *  2) points to the start of an allocated piece of memory
-     *  3) is not on a free list
-     */
-    void check(void *p) nothrow
+    GCStats stats() nothrow
     {
-        if (!p)
-        {
-            return;
-        }
+        GCStats ret;
 
-        return runLocked!(checkNoSync, otherTime, numOthers)(p);
-    }
+        runLocked!(getStatsNoSync, otherTime, numOthers)(ret);
 
+        return ret;
+    }
 
     //
     //
     //
-    private void checkNoSync(void *p) nothrow
+    private void getStatsNoSync(out GCStats stats) nothrow
     {
-        assert(p);
+        size_t psize = 0;
+        size_t usize = 0;
+        size_t flsize = 0;
 
-        sentinel_Invariant(p);
-        debug (PTRCHECK)
-        {
-            Pool*  pool;
-            size_t pagenum;
-            Bins   bin;
-            size_t size;
+        size_t n;
+        size_t bsize = 0;
 
-            p = sentinel_sub(p);
-            pool = gcx.findPool(p);
-            assert(pool);
-            pagenum = pool.pagenumOf(p);
-            bin = cast(Bins)pool.pagetable[pagenum];
-            assert(bin <= B_PAGE);
-            size = binsize[bin];
-            assert((cast(size_t)p & (size - 1)) == 0);
+        //debug(PRINTF) printf("getStats()\n");
+        memset(&stats, 0, GCStats.sizeof);
 
-            debug (PTRCHECK2)
+        for (n = 0; n < gcx.npools; n++)
+        {   Pool *pool = gcx.pooltable[n];
+
+            psize += pool.npages * PAGESIZE;
+            for (size_t j = 0; j < pool.npages; j++)
             {
-                if (bin < B_PAGE)
-                {
-                    // Check that p is not on a free list
-                    List *list;
+                Bins bin = cast(Bins)pool.pagetable[j];
+                if (bin == B_FREE)
+                    stats.freeblocks++;
+                else if (bin == B_PAGE)
+                    stats.pageblocks++;
+                else if (bin < B_PAGE)
+                    bsize += PAGESIZE;
+            }
+        }
 
-                    for (list = gcx.bucket[bin]; list; list = list.next)
-                    {
-                        assert(cast(void*)list != p);
-                    }
-                }
+        for (n = 0; n < B_PAGE; n++)
+        {
+            //debug(PRINTF) printf("bin %d\n", n);
+            for (List *list = gcx.bucket[n]; list; list = list.next)
+            {
+                //debug(PRINTF) printf("\tlist %p\n", list);
+                flsize += binsize[n];
             }
         }
-    }
 
+        usize = bsize - flsize;
 
-    /**
-     * add p to list of roots
-     */
-    void addRoot(void *p) nothrow @nogc
+        stats.poolsize = psize;
+        stats.usedsize = bsize - flsize;
+        stats.freelistsize = flsize;
+    }
+
+    void addRoot(void* p) nothrow
```

Comment:
BTW: I don't think it was a good idea to add @nogc to these functions, because it doesn't allow using garbage collected memory to store the roots, i.e. you cannot easily implement this (mostly) lock free (due to the complications freeing memory).